lib/logstash/outputs/loggly.rb

```diff
@@ -88,31 +99,78 @@ class LogStash::Outputs::Loggly < LogStash::Outputs::Base
 
   public
   def register
-    # nothing to do
   end
 
   public
+  def multi_receive(events)
+    send_batch events.collect { |event| prepare_tuple(event) }
+  end
+
   def receive(event)
+    send_batch [prepare_tuple(event)]
+  end
+
+  private
+  # Returns one event tuple [event,key,tag] or nil if event can't be sent to Loggly
+  def prepare_tuple(event)
     key = event.sprintf(@key)
     tag = event.sprintf(@tag)
 
+    if expected_field = key[/%{(.*)}/, 1]
+      @logger.warn "Skipping sending message to Loggly. No key provided (key='#{key}'). Make sure to set field '#{expected_field}'."
+      @logger.debug "Dropped message", :event => event.to_json
+      return nil
+    end
+
     # For those cases where %{somefield} doesn't exist
     # we should ship logs with the default tag value.
-    tag = 'logstash' if /^%{\w+}/.match(tag)
+    tag = DEFAULT_LOGGLY_TAG if /%{\w+}/.match(tag)
 
-    # Send event
-    send_event("#{@proto}://#{@host}/inputs/#{key}/tag/#{tag}", format_message(event))
-  end # def receive
+    [event.to_hash, key, tag]
```

Comment:
I see here we're calling event.to_hash only to call `event.to_json` on `format_message`, so I think we can use proper events in the triple and let format_message convert directly from LogStash::Event to json

Reply:
I convert to hash here for testability, unfortunately. I have been hitting issues when comparing JSON objects that were equivalent, with content ordered differently.

But now that I spell it out, I guess a better approach would be to go to the length of decoding the JSON in the tests instead. The current approach is adding work in the main flow of the plugin for the purpose of tests, which is not ideal.